Plugins/Public/dronebays/Utility.cpp

```diff
@@ -0,0 +1,343 @@
+	while((pd = Players.traverse_active(pd)))
+	{
+		HkChangeIDSString(pd->iOnlineID, currInfocard, droneName);
+	}
+
+	// Return the changed infocard and increment the counter
+	return currInfocard++;
+}
+
+void Utility::LogEvent(const char* szString, ...)
+{
+	char szBufString[1024];
+	va_list marker;
+	va_start(marker, szString);
+	_vsnprintf(szBufString, sizeof(szBufString) - 1, szString, marker);
+
+	char szBuf[64];
+	time_t tNow = time(0);
+	struct tm *t = localtime(&tNow);
+	strftime(szBuf, sizeof(szBuf), "%d/%m/%Y %H:%M:%S", t);
+
+	FILE *Logfile = fopen(("./flhook_logs/flhook_drones.log"), "at");
+	if (Logfile)
+	{
+		fprintf(Logfile, "%s %s\n", szBuf, szBufString);
+		fflush(Logfile);
+		fclose(Logfile);
+	}
+}
+
+void Utility::DeployDrone(uint iClientID, const DroneBuildTimerWrapper& timerWrapper)
+{
+	// Set the users client state to reflect a drone has been deployed
+	clientDroneInfo[iClientID].buildState = STATE_DRONE_LAUNCHED;
+
+	// Get the current system and location of the carrier
+	uint iShip;
+	Vector shipPos{};
+	Matrix shipRot{};
+	uint shipSys;
+
+	pub::Player::GetShip(iClientID, iShip);
+	pub::SpaceObj::GetLocation(iShip, shipPos, shipRot);
+	pub::SpaceObj::GetSystem(iShip, shipSys);
+
+	// If the ship doesn't exist, the carrier has docked. This should already be taken care of by the docking hook, but abort anyways.
+	if (!iShip)
+	{
+		PrintUserCmdText(iClientID, L"Info [DroneBays] :: You've docked and somehow managed to skip a Hook. Contact a dev-team FLHooker about this bug please!");
+		return;
+	}
+
+	Utility::CreateNPC(iClientID, shipPos, shipRot, shipSys, timerWrapper.reqDrone);
+	PrintUserCmdText(iClientID, L"Info :: Drone Launched");
+	
+	// Log event
+	const wstring charname = reinterpret_cast<const wchar_t*>(Players.GetActiveCharacterName(iClientID));
+	wstring logString = L"Player %s launched drone";
+	logString = ReplaceStr(logString, L"%s", charname);
+	LogEvent(wstos(logString).c_str());
+
+	// Print launch event to surrounding players
+	wstring wscMsg = L"%user has launched a drone";
+	wscMsg = ReplaceStr(wscMsg, L"%user", charname);
+	PrintLocalUserCmdText(iClientID, wscMsg, 10000);
+
+}
+
+float Utility::RandFloatRange(float a, float b)
+{
+	return ((b - a)*(static_cast<float>(rand()) / RAND_MAX)) + a;
+}
+
+void Utility::CreateNPC(uint iClientID, Vector pos, Matrix rot, uint iSystem, DroneArch drone)
+{
+
+	pub::SpaceObj::ShipInfo si;
+	memset(&si, 0, sizeof(si));
+	si.iFlag = 1;
+	si.iSystem = iSystem;
+	si.iShipArchetype = drone.archetype;
+	si.vPos = pos;
+	si.vPos.x = pos.x + RandFloatRange(0, 500);
+	si.vPos.y = pos.y + RandFloatRange(0, 500);
+	si.vPos.z = pos.z + RandFloatRange(0, 1000);
+	si.mOrientation = rot;
+	si.iLoadout = drone.loadout;
+	si.iLook1 = CreateID("li_newscaster_head_gen_hat");
+	si.iLook2 = CreateID("pl_female1_journeyman_body");
+	si.iComm = CreateID("comm_br_darcy_female");
+	si.iPilotVoice = CreateID("pilot_f_leg_f01a");
+	si.iHealth = -1;
+	si.iLevel = 23;

```

Comment:
arbitrary?

Reply:
System needs something in order for it to work. Arbitrary. I could do a level calculation, but it's not really going to add anything to the features.

Reply:
ok